src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -118,14 +110,55 @@ public async Task StartContainerAsync(IExecutionContext executionContext, object
                 }
             }
 
-#if !OS_WINDOWS
             executionContext.Debug($"Delete stale container networks from previous jobs");
             int networkPruneExitCode = await _dockerManger.DockerNetworkPrune(executionContext);
             if (networkPruneExitCode != 0)
             {
                 executionContext.Warning($"Delete stale container networks failed, docker network prune fail with exit code {networkPruneExitCode}");
             }
-#endif
+
+            // Create local docker network for this job to avoid port conflict when multiple agents run on same machine.
+            // All containers within a job join the same network
+            var containerNetwork = $"vsts_network_{Guid.NewGuid().ToString("N")}";
+            await CreateContainerNetworkAsync(executionContext, containerNetwork);
+            containers.ForEach(container => container.ContainerNetwork = containerNetwork);
+
+            foreach (var container in containers)
+            {
+                await StartContainerAsync(executionContext, container);
+            }
+        }
+
+        public async Task StopContainersAsync(IExecutionContext executionContext, object data)
+        {
+            Trace.Entering();
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+
+            List<ContainerInfo> containers = data as List<ContainerInfo>;
+            ArgUtil.NotNull(containers, nameof(containers));
+
+            var uniqueNetworks = new HashSet<string>();
+            var containerNetwork = containers.FirstOrDefault().ContainerNetwork;
+            foreach (var container in containers)
+            {
+                await StopContainerAsync(executionContext, container);
+            }
+            // Remove the container network
+            await RemoveContainerNetworkAsync(executionContext, containerNetwork);
+        }
+
+        private async Task StartContainerAsync(IExecutionContext executionContext, ContainerInfo container)
+        {
+            Trace.Entering();
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNull(container, nameof(container));
+            ArgUtil.NotNullOrEmpty(container.ContainerImage, nameof(container.ContainerImage));
+
+            Trace.Info($"Container name: {container.ContainerName}");
```

Comment:
you might want trace out those property that you newly added. port/volume ?

Reply:
Traced ports/volumes